src/NuGet.Clients/NuGet.PackageManagement.UI/Utility/PackageSourceMappingUtility.cs

```diff
@@ -91,5 +91,52 @@ internal static class PackageSourceMappingUtility
 
             mappingProvider.SavePackageSourceMappings(newAndExistingPackageSourceMappingItems);
         }
+
+        internal static string? GetNewSourceMappingSourceName(PackageSourceMapping packageSourceMapping, PackageSourceMoniker activePackageSourceMoniker)
+        {
+            string? sourceMappingSourceName = packageSourceMapping.IsEnabled
+                && !activePackageSourceMoniker.IsAggregateSource
+                ? activePackageSourceMoniker.PackageSourceNames.First() : null;
+
+            return sourceMappingSourceName;
+        }
+
+        internal static void GetNewSourceMappingsFromAddedPackages(ref Dictionary<string, SortedSet<string>>? newSourceMappings, string newMappingSourceName, List<AccessiblePackageIdentity> added, PackageSourceMapping packageSourceMapping)
+        {
+            if (newMappingSourceName is null || added.Count == 0 || packageSourceMapping is null)
+            {
+                return;
+            }
+
+            List<string> addedPackagesWithNoSourceMappings = added.Select(_ => _.Id)
+                .Where(addedPackage =>
+                {
+                    IReadOnlyList<string> configuredSources = packageSourceMapping.GetConfiguredPackageSources(addedPackage);
+                    return configuredSources == null || configuredSources.Count == 0;
+                })
+                .Distinct()
+                .ToList();
```

Comment:
How many collections are allocated here? It looks like n of added packages (configuredSources), then the call to Distinct() would create a new collection, as well as the call to ToList(). Could we save allocations by using a for loop and AddRange() instead?

Reply:
>  then the call to Distinct() would create a new collection

That's not how I understand `Distinct()` to work. It operates on an `IEnumerable`, and does not allocate anything, but simply tells the query to do it once iterated.

> as well as the call to ToList().

This is 1 allocation (there may be some internal temp arrays necessary as the collection grows, but that would be the same for `AddRange`, I believe as well).

> Could we save allocations by using a for loop and AddRange() instead?

Are you talking about this PR or are you talking about implementation of `GetConfiguredPackageSources()`? 
If you're talking about [GetConfiguredPackageSources](https://github.com/NuGet/NuGet.Client/blob/65e74ebcf721c07672394c598dccea2d86a4479d/src/NuGet.Core/NuGet.Configuration/PackageSourceMapping/PackageSourceMapping.cs#L38), that implementation works on the [SearchTree](https://github.com/NuGet/NuGet.Client/blob/65e74ebcf721c07672394c598dccea2d86a4479d/src/NuGet.Core/NuGet.Configuration/PackageSourceMapping/SearchTree.cs) by looking for term matches in the linked list in [SearchNodeByTerm](https://github.com/NuGet/NuGet.Client/blob/65e74ebcf721c07672394c598dccea2d86a4479d/src/NuGet.Core/NuGet.Configuration/PackageSourceMapping/SearchTree.cs#L108), which as far as I can tell doesn't allocate any collections as part of its search.

Reply:
> That's not how I understand Distinct() to work. It operates on an IEnumerable, and does not allocate anything, but simply tells the query to do it once iterated.

I'm not sure how any algorithm could determine a distinct list of items without any allocations. Here's the .NET 8 implementation: https://github.com/dotnet/runtime/blob/a39262bd2d8543e5e6636978d54a084669e1a8e5/src/libraries/System.Linq/src/System/Linq/Distinct.cs#L62-L79
It allocates a `HashSet<T>`, which itself has an internal collection that might grow depending on the number of insertions.

On .NET Framework, the method also uses `yield return`, meaning calling the method creates a finite state machine on the heap: https://referencesource.microsoft.com/#System.Core/System/Linq/Enumerable.cs,836

> This is 1 allocation (there may be some internal temp arrays necessary as the collection grows, but that would be the same for AddRange, I believe as well).

You can avoid temporary allocations (resizing) if it's possible to pre-compute the expected result list size. Then you use the constructor where you pass the capacity as a parameter. Then you can use `AddRange` (although using `IEnumerable<T>` causes allocations), so a `for` loop (or a `foreach` loop, but only when the collection you're iterating has a `struct` enumerator) can avoid all temporary allocations.

Even when you can't calculate the expected result size accurately, if you can estimate it and be close enough most of the time, it's still going to be a win, because `List<T>()`'s default constructor creates a list with zero capacity, the first add would allocate a capacity of 4, then it would double in size each time it got full. So, if you know you're going to have ballpark 10 items, using the default list capacity will cause 2 resizes. I'm not sure if newer versions of .NET have changed this, but VS runs on the .NET Framework anyway.

Having written all that, I haven't yet looked at this PR's code, I'm only writing in general terms (since "the same for AddRange" is also a general statement, not specific to this PR). For our most performance sensitive code, like restore, it's worthwhile to write extra lines of code to reduce allocations and increase performance, even when it increases code complexity by a "reasonable" amount (which unfortunately is subjective). However, for non-perf critical code, it's not worthwhile making the code a lot more complex.

I strongly encourage everyone to always practise basic performance best-practises, even in code paths that are not at all perf sensitive, so that when you later work on something that actually is on a hot path and is performance sensitive, you're already well practised in thinking about how to write code that reduces allocations, or is otherwise "high performance".

Reply:
ok, having looked though this PR, and better understanding what this method does, I have a few ideas.

One option is to create the list with a suitable capacity, and then use `AddRange`, as Arturo suggested. The upper bound of how big the result will be is `added.Count`, assuming none of the new packages have a mapping and all package IDs are unique. While it might feel like it's a waste of memory to create an array (`List<T>`'s backing store is ultimately an array), since the `addedPackagesWithNoSourceMappings` is not used outside of this method, it should hopefully be cleaned up in gen0 GC, so not overly expensive. I'm sure that would be less customer impact than requiring multiple allocations when the customer installs a package like `Microsoft.VisualStudio.SDK` and none of the packages have a mapping yet.

However, a second option is to avoid using the temporary `addedPackagesWithNoSourceMappings` list. Use a `foreach` loop on `added` (which is of type `List<T>` and therefore `foreach` is zero allocation), and then add to the dictionary directly. It means the `if (newSourceMappings is null)` check will happen multiple times, but that will have less perf impact that creating a temporary list (especially if it ever gets resized), which then needs to get GC'd.

To be honest, I'm not sure why the `Distinct()` is used in the LINQ expression. Isn't the list of `added` packages already unique? If not, that's very unintuitive to me, even in multi-targeting projects. But in that case, one option is to use your own `HashSet<T>`, as `Distinct()` does. However, [`SortedSet<T>.Add`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.sortedset-1.add?view=net-7.0) does not throw an exception when a duplicate is added. So, I don't see any perf benefit to calculating duplicates in this method. Only using `newSourceMappings.Add` achieves the same as using a `HashSet` (either explicitly, or through `.Distinct()`), except without the allocations.

So, I believe this method could avoid all temporary allocations.

Reply:
Thanks for the suggestion! @zivkan, I believe I've implemented your "second option" above. I added a single test case for the method in `PackageSourceMappingUtility` before changing it. Let me know if this is looks like what you meant. 

Looking back at my use of `Distinct`, it must have been a relic from before I started using a `SortedSet`, as I can't figure out why I'd need it. 

> Isn't the list of added packages already unique? If not, that's very unintuitive to me, even in multi-targeting projects.

I believe when I saw the loop for project actions here, I was thinking that when multiple install actions occur, which can be the same package to multiple projects, that I could see the same top-level or transitive package in a single call to `GetPreviewResultsAsync` . See loops here: https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs#L945